### PR TITLE
script: Lint Gitian descriptors with ShellCheck

### PR DESCRIPTION
**Pull request by hebasto**
This PR extracts shell scripts from Gitian descriptors (`contrib/gitian-descriptors/`) and checks for ShellCheck warnings as any other one.

Some non-controversial warnings are fixed.

### CODE REVIEW

File: test/lint/print-gitian-script.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env python3
```
Comment (promag) [inline test/lint/print-gitian-script.py]: Is this really necessary? Couldn't use some YAML processor like `yq`?

**Review by promag — COMMENTED** [review]
Concept ACK.

File: test/lint/print-gitian-script.py (hunk 1 above)
Comment (hebasto) [reply test/lint/print-gitian-script.py]: There is no straightforward [way](https://mikefarah.github.io/yq/) to install `yq` on Travis, unfortunately.

Comment (hebasto) [reply test/lint/print-gitian-script.py]: > Couldn't use some YAML processor like `yq`?

Done.
TIL, thank you ;)

**Comment by practicalswift** [comment]
Concept ACK: good idea!

Could perhaps check for `yq` and skip the Gitian checking if it is not installed (with a warning message) in order to not break current setups?

**Comment by hebasto** [comment]
@practicalswift 
> Could perhaps check for `yq` and skip the Gitian checking if it is not installed (with a warning message) in order to not break current setups?

Done. Thank you.

File: contrib/gitian-descriptors/gitian-win.yml [hunk 2]
```diff
@@ -111,10 +112,11 @@ script: |
 
   # Create the release tarball using (arbitrarily) the first host
   ./autogen.sh
-  CONFIG_SITE=${BASEPREFIX}/`echo "${HOSTS}" | awk '{print $1;}'`/share/config.site ./configure --prefix=/
+  CONFIG_SITE=${BASEPREFIX}/$(echo "${HOSTS}" | awk '{print $1;}')/share/config.site ./configure --prefix=/
   make dist
-  SOURCEDIST=`echo bitcoin-*.tar.gz`
-  DISTNAME=`echo ${SOURCEDIST} | sed 's/.tar.*//'`
+  SOURCEDIST=$(echo bitcoin-*.tar.gz)
+  DISTNAME=${SOURCEDIST//.tar.*/}
```
Comment (dongcarl) [inline contrib/gitian-descriptors/gitian-win.yml]:
```suggestion
  DISTNAME=${SOURCEDIST/%.tar.gz}
```

File: contrib/gitian-descriptors/gitian-win-signer.yml [hunk 3]
```diff
@@ -36,7 +36,7 @@ script: |
   ./configure --without-gsf --without-curl --disable-dependency-tracking
   make
   find ${UNSIGNED_DIR} -name "*-unsigned.exe" | while read i; do
-    INFILE="`basename "${i}"`"
-    OUTFILE="`echo "${INFILE}" | sed s/-unsigned//`"
+    INFILE="$(basename "${i}")"
+    OUTFILE="${INFILE//-unsigned/}"
```
Comment (dongcarl) [inline contrib/gitian-descriptors/gitian-win-signer.yml]:
```suggestion
    OUTFILE="${INFILE/%-unsigned}"
```

File: contrib/gitian-descriptors/gitian-osx.yml [hunk 4]
```diff
@@ -104,10 +105,10 @@ script: |
 
   # Create the release tarball using (arbitrarily) the first host
   ./autogen.sh
-  CONFIG_SITE=${BASEPREFIX}/`echo "${HOSTS}" | awk '{print $1;}'`/share/config.site ./configure --prefix=/
+  CONFIG_SITE=${BASEPREFIX}/$(echo "${HOSTS}" | awk '{print $1;}')/share/config.site ./configure --prefix=/
   make dist
-  SOURCEDIST=`echo bitcoin-*.tar.gz`
-  DISTNAME=`echo ${SOURCEDIST} | sed 's/.tar.*//'`
+  SOURCEDIST=$(echo bitcoin-*.tar.gz)
+  DISTNAME=${SOURCEDIST//.tar.*/}
```
Comment (dongcarl) [inline contrib/gitian-descriptors/gitian-osx.yml]: Same here.

File: contrib/gitian-descriptors/gitian-linux.yml [hunk 5]
```diff
@@ -141,10 +142,11 @@ script: |
 
   # Create the release tarball using (arbitrarily) the first host
   ./autogen.sh
-  CONFIG_SITE=${BASEPREFIX}/`echo "${HOSTS}" | awk '{print $1;}'`/share/config.site ./configure --prefix=/
+  CONFIG_SITE=${BASEPREFIX}/$(echo "${HOSTS}" | awk '{print $1;}')/share/config.site ./configure --prefix=/
   make dist
-  SOURCEDIST=`echo bitcoin-*.tar.gz`
-  DISTNAME=`echo ${SOURCEDIST} | sed 's/.tar.*//'`
+  SOURCEDIST=$(echo bitcoin-*.tar.gz)
+  DISTNAME=${SOURCEDIST//.tar.*/}
```
Comment (dongcarl) [inline contrib/gitian-descriptors/gitian-linux.yml]: Same here.

**Review by dongcarl — CHANGES_REQUESTED** [review]

**Comment by hebasto** [comment]
@dongcarl 
Thank you for your review. All your comments have been addressed.

File: contrib/gitian-descriptors/gitian-win.yml (hunk 2 above)
Comment (hebasto) [reply contrib/gitian-descriptors/gitian-win.yml]: Done.

File: contrib/gitian-descriptors/gitian-win-signer.yml (hunk 3 above)
Comment (hebasto) [reply contrib/gitian-descriptors/gitian-win-signer.yml]: Done.

File: contrib/gitian-descriptors/gitian-osx.yml (hunk 4 above)
Comment (hebasto) [reply contrib/gitian-descriptors/gitian-osx.yml]: Done.

File: contrib/gitian-descriptors/gitian-linux.yml (hunk 5 above)
Comment (hebasto) [reply contrib/gitian-descriptors/gitian-linux.yml]: Done.

File: test/lint/lint-shell.sh [hunk 6]
```diff
@@ -21,11 +21,40 @@ if ! command -v shellcheck > /dev/null; then
...
+if ! command -v yq > /dev/null; then
+    echo "Skipping Gitian desriptor scripts checking since yq is not installed."
+    exit $EXIT_CODE
+fi
+
+for descriptor in $(git ls-files -- 'contrib/gitian-descriptors/*.yml')
+do
+    echo
+    echo "$descriptor"
+    SCRIPT=$'#!/bin/bash\n'$(yq -r .script "$descriptor")
```
Comment (dongcarl) [inline test/lint/lint-shell.sh]:
```suggestion
    SCRIPT=$'#!/usr/bin/env bash\n'$(yq -r .script "$descriptor")
```

**Review by dongcarl — CHANGES_REQUESTED** [review]
For all the places where you had to split `export DIR=$(pwd)`, just use `export DIR="$PWD"`, which doesn't require the splitting

File: test/lint/lint-shell.sh (hunk 6 above)
Comment (hebasto) [reply test/lint/lint-shell.sh]: It just mimics the way [gitian-builder/bin/gbuild](https://github.com/devrandom/gitian-builder/blob/cd8aa53c4e8870fa6be614a653024bae094c2f22/bin/gbuild#L128) makes shell script:
```
  File.open("var/build-script", "w") do |script|
    script.puts "#!/bin/bash"
```

Comment (dongcarl) [reply test/lint/lint-shell.sh]: Ah! Makes sense to do it this way, could you add a comment?

Comment (hebasto) [reply test/lint/lint-shell.sh]: Sure!

Comment (hebasto) [reply test/lint/lint-shell.sh]: Done.

**Comment by hebasto** [comment]
> For all the places where you had to split `export DIR=$(pwd)`, just use `export DIR="$PWD"`, which doesn't require the splitting

Done.

**Comment by laanwj** [comment]
Concept ACK

**Review by promag — COMMENTED** [review]
Please update https://github.com/bitcoin/bitcoin/blob/master/test/README.md#dependencies-1

**Comment by hebasto** [comment]
@promag 
> Please update https://github.com/bitcoin/bitcoin/blob/master/test/README.md#dependencies-1

It already lacks `ShellCheck`. Maybe just add `yq` dependency to #17353?

**Comment by promag** [comment]
@hebasto let's see if that one is merged first so you can update the doc here.

**Comment by hebasto** [comment]
Latest push:
- rebased on top of merged #17353 and @promag's https://github.com/bitcoin/bitcoin/pull/17361#pullrequestreview-312342839 has been addressed
- @laanwj's https://github.com/bitcoin/bitcoin/pull/17353#discussion_r343037349 have been addresed

**Comment by hebasto** [comment]
@dongcarl @promag 

Thank you for your reviews. All your comments have been addressed.
Would you mind re-reviewing this PR?

**Comment by dongcarl** [comment]
ACK from me if Gitian build passes.

**Comment by practicalswift** [comment]
ACK 93b18e5f8de0669272618e8098b7c3fdc435906b -- diff looks correct

File: contrib/gitian-descriptors/gitian-win.yml [hunk 7]
```diff
@@ -96,7 +96,7 @@ script: |
   export PATH=${WRAP_DIR}:${PATH}
 
   cd bitcoin
-  BASEPREFIX=`pwd`/depends
+  BASEPREFIX=$(pwd)/depends
```
Comment (laanwj) [inline contrib/gitian-descriptors/gitian-win.yml]: Did you mean `${PWD}`?

File: contrib/gitian-descriptors/gitian-linux.yml [hunk 8]
```diff
@@ -122,7 +122,7 @@ script: |
   done
 
   cd bitcoin
-  BASEPREFIX=`pwd`/depends
+  BASEPREFIX=$(pwd)/depends
```
Comment (dongcarl) [inline contrib/gitian-descriptors/gitian-linux.yml]: Let's be consistent and just use `$PWD` everywhere. 
```suggestion
  BASEPREFIX="${PWD}/depends"
```

Comment (hebasto) [reply contrib/gitian-descriptors/gitian-linux.yml]: > Let's be consistent and just use `$PWD` everywhere.

Agree.

Ref:
```bash
$ echo $0
bash
$ type pwd
pwd is a shell builtin
$ pwd --help
pwd: pwd [-LP]
    Print the name of the current working directory.
    
    Options:
      -L	print the value of $PWD if it names the current working
    		directory
      -P	print the physical directory, without any symbolic links
    
    By default, `pwd' behaves as if `-L' were specified.
    
    Exit Status:
    Returns 0 unless an invalid option is given or the current directory
    cannot be read.
```

Going to fix tonight.

Comment (hebasto) [reply contrib/gitian-descriptors/gitian-linux.yml]: Done.

**Comment by hebasto** [comment]
All comments has been addressed.

**Comment by practicalswift** [comment]
ACK 17f81e96486780df5d464487975ecb11b278ec8d -- diff looks correct

**Pull request merged**